File: java-wrapper/src/main/java/com/codedisaster/steamworks/SteamMatchmakingGameServerItem.java
package com.codedisaster.steamworks;

public class SteamMatchmakingGameServerItem {

	SteamMatchmakingServerNetAdr netAdr = new SteamMatchmakingServerNetAdr();
	int ping;
	boolean hadSuccessfulResponse;
	boolean doNotRefresh;
	String gameDir;
	String map;
	String gameDescription;
	int appID;
	int players;
	int maxPlayers;
	int botPlayers;
	boolean password;
	boolean secure;
	int timeLastPlayed;
	int	serverVersion;

	String serverName;

	String gameTags;

	long steamID;

	public SteamMatchmakingGameServerItem() {

	}

	public SteamMatchmakingServerNetAdr getNetAdr() {
		return netAdr;
	}

	public int getPing() {
		return ping;
	}

	public boolean hadSuccessfulResponse() {
		return hadSuccessfulResponse;
	}

	public boolean doNotRefresh() {
		return doNotRefresh;
	}

	public String getGameDir() {
		return gameDir;
	}

	public String getMap() {
		return map;
	}

	public String getGameDescription() {
		return gameDescription;
	}

	public int getAppID() {
		return appID;
	}

	public int getPlayers() {
		return players;
	}

	public int getMaxPlayers() {
		return maxPlayers;
	}

	public int getBotPlayers() {
		return botPlayers;
	}

	public boolean hasPassword() {
		return password;
	}

	public boolean isSecure() {
		return secure;
	}

	public int getTimeLastPlayed() {
		return timeLastPlayed;
	}

	public int getServerVersion() {
		return serverVersion;
	}

	public String getServerName() {
		return serverName;
	}

	public String getGameTags() {
		return gameTags;
	}

	public SteamID getSteamID() {
		return new SteamID(steamID);
	}

}

File: java-wrapper/src/main/native/SteamMatchmakingGameServerItem.cpp
#include "SteamMatchmakingGameServerItem.h"

void convertGameServerItem(jobject item, JNIEnv* env, const gameserveritem_t& server) {
	jclass ex = env->FindClass("com/codedisaster/steamworks/SteamException");
	jclass itemClazz = env->FindClass("com/codedisaster/steamworks/SteamMatchmakingGameServerItem");

	jfieldID field = env->GetFieldID(itemClazz, "netAdr", "Lcom/codedisaster/steamworks/SteamMatchmakingServerNetAdr;");
	jclass netAdrClazz = env->FindClass("com/codedisaster/steamworks/SteamMatchmakingServerNetAdr");
	jobject netAdr = env->GetObjectField(item, field);
	if (netAdrClazz != nullptr && netAdr != nullptr) {
		field = env->GetFieldID(netAdrClazz, "connectionPort", "S");
		env->SetShortField(netAdr, field, server.m_NetAdr.GetConnectionPort());

		field = env->GetFieldID(netAdrClazz, "queryPort", "S");
		env->SetShortField(netAdr, field, server.m_NetAdr.GetQueryPort());

		field = env->GetFieldID(netAdrClazz, "ip", "I");
		env->SetIntField(netAdr, field, server.m_NetAdr.GetIP());
	} else {
		env->ThrowNew(ex, "Could not access field: SteamMatchmakingGameServerItem#netAdr");
	}

	field = env->GetFieldID(itemClazz, "ping", "I");
	env->SetIntField(item, field, server.m_nPing);

	field = env->GetFieldID(itemClazz, "hadSuccessfulResponse", "Z");
	env->SetBooleanField(item, field, server.m_bHadSuccessfulResponse);

	field = env->GetFieldID(itemClazz, "doNotRefresh", "Z");
	env->SetBooleanField(item, field, server.m_bDoNotRefresh);

	field = env->GetFieldID(itemClazz, "gameDir", "Ljava/lang/String;");
	env->SetObjectField(item, field, env->NewStringUTF(server.m_szGameDir));

	field = env->GetFieldID(itemClazz, "map", "Ljava/lang/String;");
	env->SetObjectField(item, field, env->NewStringUTF(server.m_szMap));

	field = env->GetFieldID(itemClazz, "gameDescription", "Ljava/lang/String;");
	env->SetObjectField(item, field, env->NewStringUTF(server.m_szGameDescription));

	field = env->GetFieldID(itemClazz, "appID", "I");
	env->SetIntField(item, field, server.m_nAppID);

	field = env->GetFieldID(itemClazz, "players", "I");
	env->SetIntField(item, field, server.m_nPlayers);

	field = env->GetFieldID(itemClazz, "maxPlayers", "I");
	env->SetIntField(item, field, server.m_nMaxPlayers);

	field = env->GetFieldID(itemClazz, "botPlayers", "I");
	env->SetIntField(item, field, server.m_nBotPlayers);

	field = env->GetFieldID(itemClazz, "password", "Z");
	env->SetBooleanField(item, field, server.m_bPassword);

	field = env->GetFieldID(itemClazz, "secure", "Z");
	env->SetBooleanField(item, field, server.m_bSecure);

	field = env->GetFieldID(itemClazz, "timeLastPlayed", "I");
	env->SetIntField(item, field, server.m_ulTimeLastPlayed);

	field = env->GetFieldID(itemClazz, "serverVersion", "I");
	env->SetIntField(item, field, server.m_nServerVersion);

	field = env->GetFieldID(itemClazz, "serverName", "Ljava/lang/String;");
	env->SetObjectField(item, field, env->NewStringUTF(server.GetName()));

	field = env->GetFieldID(itemClazz, "gameTags", "Ljava/lang/String;");
	env->SetObjectField(item, field, env->NewStringUTF(server.m_szGameTags));

	field = env->GetFieldID(itemClazz, "steamID", "J");
	env->SetLongField(item, field, server.m_steamID.ConvertToUint64());
}

jobject createGameServerItem(JNIEnv* env, const gameserveritem_t& server) {

	jclass ex = env->FindClass("com/codedisaster/steamworks/SteamException");
	jclass itemClazz = env->FindClass("com/codedisaster/steamworks/SteamMatchmakingGameServerItem");

	if (itemClazz != nullptr) {
		jmethodID itemCtor = env->GetMethodID(itemClazz, "<init>", "()V");
		if (itemCtor != nullptr) {
			jobject item = env->NewObject(itemClazz, itemCtor);
			convertGameServerItem(item, env, server);
			return item;
		} else {
			env->ThrowNew(ex, "Could not create instance of class: SteamMatchmakingGameServerItem");
		}
	} else {
		env->ThrowNew(ex, "Could not find class: SteamMatchmakingGameServerItem");
	}

	return nullptr;
}
